search/controllers/api/classic_parser.py

```diff
@@ -0,0 +1,190 @@
+    """Cast tokens to class-based representations."""
+    classed_tokens: List[Union[Operator, Term, Phrase]] = []
+    for token in tokens:
+        if isinstance(token, str):
+            if ':' in token:
+                token = _parse_field_query(token)
+                classed_tokens.append(token)
+            else:
+                operator = _parse_operator(token)
+                classed_tokens.append(operator)
+        else:
+            classed_tokens.append(token)
+
+    return classed_tokens
+
+
+def _group_tokens(classed_tokens: List[Union[Operator, Term, Phrase]]) -> Phrase:
+    """Group operators together with Term."""
+    phrases: List[Phrase] = []
+    current_op: Optional[Operator] = None
+    prev_token: Optional[Union[Operator, Term, Phrase]] = None
+    for token in classed_tokens:
+        if isinstance(token, Operator) and isinstance(prev_token, Operator):
+            raise BadRequest(f'Query contains 2 consecutive operators: {prev_token} {token}')
+        if isinstance(token, Operator):
+            current_op = token
+        else:
+            if current_op:
+                token = (current_op, token)
+            phrases.append(token)
+        prev_token = token
+
+    # Return single-token query, otherwise cast to a tuple.
+    if len(phrases) == 1:
+        return phrases[0]
+    else:
+        # When mypy adds full support for recursive types, this should be fine.
+        return tuple(phrases)  # type: ignore
+
+def _parse_operator(characters: str) -> Operator:
+    try:
+        return Operator(characters.strip())
+    except ValueError as e:
+        raise BadRequest(f'Cannot parse fragment: {characters}') from e
+
+
+def _parse_field_query(field_part: str) -> Term:
+    field_name, value = field_part.split(':', 1)
+
+    # cast field to Field enum
+    try:
+        field = Field(field_name)
+    except ValueError as e:
+        raise BadRequest(f'Invalid field: {field_name}') from e
+
+    # process quotes, if present
+    if value.startswith('"') and value.endswith('"'):
+        value = value[1:-1]
+
+    return field, value
+
+
+def phrase_to_query_string(phrase: Phrase) -> str:
+    """Converts a phrase object back to a query string."""
+    parts: List[str] = []
+    for token in phrase:
+        if isinstance(token, Operator):
+            parts.append(token.value)
+        elif isinstance(token, Field):
+            parts.append(f"{token.value}:")
+        elif isinstance(token, str):
+            # strings are added to the field or operator preceeding it 
+            if ' ' in token:
+                parts[-1] += f'"{token}"'
+            else:
+                parts[-1] += token
+        elif isinstance(token, tuple):
+            part = phrase_to_query_string(token)
+            # If the returned part is a Phrase, add parens
+            if ' ' in part and part.count(':') > 1 \
+                and part.split()[0] not in map(attrgetter('value'), Operator):  # doesn't start with an operator
```

Comment:
```suggestion
                    and part.split()[0] not in map(attrgetter('value'), Operator):  # Doesn't start with an operator.
```
Need another four spaces for correct continuation. Double-check that the line is not too long with that inline comment, though.